cdap-api/src/main/java/co/cask/cdap/api/spark/AbstractSpark.java

```diff
@@ -110,12 +113,43 @@ protected final void setExecutorResources(Resources resources) {
   }
 
   @Override
+  @Deprecated
   public void beforeSubmit(SparkClientContext context) throws Exception {
     // Do nothing by default
   }
 
   @Override
+  @Deprecated
   public void onFinish(boolean succeeded, SparkClientContext context) throws Exception {
     // Do nothing by default
   }
+
+  @Override
+  public final void initialize(SparkClientContext context) throws Exception {
+    this.context = context;
+    initialize();
+  }
+
+  /**
+   * Classes derived from {@link AbstractSpark} can override this method to initialize the {@link Spark}.
+   * {@link SparkClientContext} will be available in this method using {@link AbstractSpark#getContext}.
+   * Default implementation of this method calls the deprecated {@link AbstractSpark#beforeSubmit} method.
+   * @throws Exception if there is any error in initializing the MapReduce
+   */
+  protected void initialize() throws Exception {
+    beforeSubmit(context);
+  }
+
+
+  @Override
+  public void destroy() {
+    //do nothing by default
```

Comment:
Shouldn't it call `onFinish` by default, similar to what `AbstractMapReduce` does?